src/spam-protection/spam-protection-action-creator.ts

```diff
@@ -48,17 +44,37 @@ export default class SpamProtectionActionCreator {
         );
     }
 
-    execute(): ThunkAction<SpamProtectionAction, InternalCheckoutSelectors> {
+    executeCheckoutSpamProtection(): ThunkAction<SpamProtectionAction> {
         return store => defer(() => {
-            const { checkout } = store.getState();
-            const { id: checkoutId, shouldExecuteSpamCheck } = checkout.getCheckoutOrThrow();
+            const state = store.getState();
+            const checkout = state.checkout.getCheckout();
 
-            if (!shouldExecuteSpamCheck) {
-                return empty();
+            if (!checkout) {
+                throw new MissingDataError(MissingDataErrorType.MissingCheckout);
             }
 
+            return from(this.execute()(store))
+                .pipe(last())
```

Comment:
Similarly, `last()` means taking the last emitted action so it'd filter out all the other actions that are emitted. Also, I'd emit `Requested` and `Failed` actions, not just `Succeeded`, because it is an async action creator. I'd do something like this instead:

```ts
return concat(
    createAction(SpamProtectionActionType.CheckoutVerifyRequested),
    from(this.execute()(store))
        .pipe(switchMap(action =>
            isSpamProtectionExecuteSucceededAction(action) && action.payload?.token ?
                from(this._requestSender.validate(checkout.id, action.payload.token))
                    .pipe(switchMap(({ body }) => [
                        action,
                        createAction(SpamProtectionActionType.CheckoutVerifySucceeded, body),
                    ])) :
                of(action)
        ))
).pipe(
    catchError(error => throwErrorAction(SpamProtectionActionType.CheckoutVerifyFailed, error))
);
```